calyx/src/passes/minimize_regs.rs

```diff
@@ -54,62 +40,70 @@ impl Visitor for MinimizeRegs {
+            if let Some(name) = cell.type_name() {
+                if name == "std_reg" {
+                    return Some((
+                        cell.get_paramter("width").unwrap(),
+                        cell.name.clone(),
+                    ));
+                }
+            }
+            None
+        });
 
-        // add constraints between things that are alive at the same time
-        for group in &comp.groups {
-            let conflicts = self.live.get(&group.borrow());
-            self.graph.insert_conflicts(
-                &conflicts.iter().cloned().collect::<Vec<_>>(),
-            );
+        let mut graph: GraphColoring<ir::Id> =
+            GraphColoring::from(registers.clone().map(|(_, name)| name));
 
-            let in_parallel: Vec<_> = self
-                .par_conflicts
-                .all_conflicts(&group)
-                .into_iter()
-                .flat_map(|group| {
-                    self.live
-                        .get(&*group.borrow())
-                        .iter()
-                        .cloned()
-                        .collect::<Vec<_>>()
-                })
-                .collect();
+        let live = LiveRangeAnalysis::new(&comp, &*comp.control.borrow());
+        let par_conflicts = ScheduleConflicts::from(&*comp.control.borrow());
 
-            for conflict_here in conflicts {
-                for par_conflict in &in_parallel {
-                    self.graph.insert_conflict(
-                        conflict_here.clone(),
-                        par_conflict.clone(),
-                    );
-                }
-            }
+        for group in &comp.groups {
+            // Conflict edges between all things alive at the same time.
+            let conflicts = live.get(&group.borrow().name);
+            graph.insert_conflicts(conflicts.iter());
         }
 
-        // add constraints so that registers of different sizes can't be shared
-        for a_ref in &comp.cells {
-            for b_ref in &comp.cells {
-                let a = a_ref.borrow();
-                let b = b_ref.borrow();
-                let a_correct_type = a.type_name() == Some(&"std_reg".into());
-                let b_correct_type = b.type_name() == Some(&"std_reg".into());
-                if !(a_correct_type && b_correct_type) {
-                    continue;
-                }
+        // Conflict edges between all groups that are enabled in parallel.
+        par_conflicts
+            .all_conflicts()
+            .into_grouping_map_by(|(g1, _)| g1.clone())
+            .fold(vec![], |mut acc, _, (_, conf_group)| {
+                acc.extend(live.get(&conf_group));
+                acc
+            })
+            .into_iter()
+            .for_each(|(group, confs)| {
+                let conflicts = live.get(&group);
+                confs
+                    .into_iter()
+                    // This unique call saves a lot of time!
+                    .unique()
```

Comment:
This call is really beneficial. I wonder why we even have so many duplicates in the first place.